### PR TITLE
Surface Curvature Feature: explicit local κ for surface nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model's input features (DSDF, TE coord frame, wake deficit) encode **position** relative to aerodynamic reference points, but carry no explicit information about **shape variation** along the foil surface. Local curvature κ = dθ/ds (rate of change of tangent angle with arc-length) directly determines the surface pressure distribution via the Euler equations — high curvature at the LE creates the suction peak, gentle curvature at mid-chord gives smooth pressure recovery, and TE curvature governs separation behavior.

Computing curvature from existing inputs requires second-order spatial derivatives of DSDF, which neural networks have strong spectral bias against learning. Providing κ explicitly as an input feature removes this hard learning task — the same reasoning that made TE coord frame (-5.4% p_in) and wake deficit (-4.1% p_in) work.

**Physical motivation:**
- Inviscid pressure: Cp ∝ local curvature (Euler equations on body-conforming coordinates)
- Boundary layer: separation onset correlates with adverse pressure gradient, which tracks curvature changes
- OOD generalization: dimensionless curvature (κ × chord) is shape-invariant — a NACA0012 and NACA6416 have different curvature distributions that directly explain their different Cp distributions

**Key distinction from in-flight experiments:**
- Surface normals (#2229 alphonse): encode **direction** the surface faces → (nx, ny)
- Surface curvature (this PR): encodes **rate of change** of direction → dθ/ds
- These are complementary: normals are first-order, curvature is second-order geometry

**Confidence:** Medium-high. Follows the proven feature engineering pattern. The only risk is that DSDF already partially encodes curvature information (level-set curvature = ∇²DSDF), but making it explicit should still help given the spectral bias argument.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flag

In the `Config` dataclass:
```python
surface_curvature_feature: bool = False   # local curvature κ for surface nodes
```

Add argparse: `--surface_curvature_feature` (store_true).

### Step 2: Implement curvature computation

Add a helper function (similar to the TE coord frame helper):

```python
def compute_surface_curvature(x_coords, boundary_id, n_input):
    """Compute dimensionless local curvature for surface nodes.
    
    Args:
        x_coords: [B, N, n_input] node features (x,y at indices 0,1)
        boundary_id: [B, N] integer boundary IDs
        n_input: number of input channels
    
    Returns:
        kappa_fore: [B, N, 1] curvature w.r.t fore-foil (0 for volume nodes)
        kappa_aft:  [B, N, 1] curvature w.r.t aft-foil (0 for volume/single-foil)
    """
    B, N, _ = x_coords.shape
    kappa_fore = torch.zeros(B, N, 1, device=x_coords.device)
    kappa_aft = torch.zeros(B, N, 1, device=x_coords.device)
    
    for foil_bid, kappa_out in [(6, kappa_fore), (7, kappa_aft)]:
        for b in range(B):
            mask = (boundary_id[b] == foil_bid)
            if mask.sum() < 3:
                continue  # need at least 3 nodes for curvature
            
            xy = x_coords[b, mask, :2]  # [M, 2]
            
            # Sort by angle from centroid (gives consistent ring ordering)
            centroid = xy.mean(dim=0, keepdim=True)
            angles = torch.atan2(xy[:, 1] - centroid[0, 1],
                                  xy[:, 0] - centroid[0, 0])
            order = angles.argsort()
            xy_sorted = xy[order]
            
            # Compute tangent vectors between consecutive nodes
            dx = xy_sorted[1:, 0] - xy_sorted[:-1, 0]
            dy = xy_sorted[1:, 1] - xy_sorted[:-1, 1]
            ds = torch.sqrt(dx**2 + dy**2 + 1e-8)
            
            # Tangent angles
            theta = torch.atan2(dy, dx)  # [M-1]
            
            # Curvature: dtheta/ds (central difference where possible)
            # For M surface nodes we get M-2 curvature values (interior nodes)
            dtheta = theta[1:] - theta[:-1]  # [M-2]
            # Handle angle wrapping
            dtheta = torch.atan2(torch.sin(dtheta), torch.cos(dtheta))
            ds_mid = 0.5 * (ds[1:] + ds[:-1])  # [M-2]
            kappa = dtheta / (ds_mid + 1e-8)  # [M-2]
            
            # Normalize by chord length (max_x - min_x) for dimensionless curvature
            chord = xy[:, 0].max() - xy[:, 0].min() + 1e-8
            kappa_norm = kappa * chord  # dimensionless
            
            # Assign back to original node positions
            # order gives sorted→original mapping. Interior sorted nodes are indices 1..M-2
            interior_sorted_idx = order[1:-1]  # indices in the original mask space
            # Map back to full [N] space
            mask_indices = mask.nonzero(as_tuple=True)[0]
            full_indices = mask_indices[interior_sorted_idx]
            kappa_out[b, full_indices, 0] = kappa_norm
            
            # Extrapolate endpoints: copy from nearest interior node
            kappa_out[b, mask_indices[order[0]], 0] = kappa_norm[0]
            kappa_out[b, mask_indices[order[-1]], 0] = kappa_norm[-1]
    
    return kappa_fore, kappa_aft
```

### Step 3: Integrate into the input pipeline

In the preprocessing section where TE coord frame and wake deficit features are computed:

```python
if cfg.surface_curvature_feature:
    kappa_fore, kappa_aft = compute_surface_curvature(x, boundary_id, n_input)
    x = torch.cat([x, kappa_fore, kappa_aft], dim=-1)
    # Update n_input accordingly (add 2 channels)
```

Make sure `n_input` is updated at model construction time too:
```python
if cfg.surface_curvature_feature:
    n_input += 2  # kappa_fore, kappa_aft
```

### Step 4: Handle all 4 loops

Apply the curvature computation in ALL loops: train, val, vis, and re-verification. The computation is per-sample and uses only mesh geometry, so it's deterministic and safe.

### Step 5: torch.compile compatibility

All operations (atan2, sort, indexing, finite differences) are compile-safe for fixed-size tensors. If the per-sample loop over `B` causes issues with torch.compile, move the computation to a pre-processing step before the training loop (compute once per sample and cache, since curvature depends only on mesh geometry, not on predictions).

**IMPORTANT optimization note:** Curvature depends only on the mesh geometry, which is FIXED per sample. You can precompute curvature for each sample once and cache it in the dataset/dataloader. This avoids recomputing every epoch. If caching is complex, the per-batch computation is acceptable since it's just finite differences on ~150 surface nodes.

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/surface-curvature-s42" \
  --wandb_group "round18/surface-curvature-feature" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --surface_curvature_feature

# Seed 73 — identical but --seed 73 --wandb_name "askeladd/surface-curvature-s73"
```

### Step 7: Report results

Add a **Results** comment to this PR with:
- Table: p_in, p_oodc, p_tan, p_re for seed 42, seed 73, and 2-seed average
- Comparison against baseline (deltas)
- W&B run IDs
- Note on where curvature values are largest (LE, TE, mid-chord) and if this correlates with prediction improvement regions

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Results — Surface Curvature Feature

### Summary

Implemented `compute_surface_curvature()` helper that computes dimensionless local curvature κ×chord for fore-foil and aft-foil surface nodes separately. Uses angle-sorted node ordering from centroid and finite differences of tangent angle (dθ/ds). Added as 2 input channels (kappa_fore, kappa_aft) — volume nodes get 0.

### Metrics

| Metric | Baseline (PR #2213) | S42 | S73 | 2-seed avg | Delta |
|--------|---------------------|-----|-----|------------|-------|
| p_in   | **11.979** | 47.9 | 14.4 | 31.15 | **+160%** |
| p_tan  | **28.341** | 65.3 | 29.8 | 47.55 | **+68%** |
| p_oodc | **7.643** | 38.7 | 9.5 | 24.10 | **+215%** |
| p_re   | **6.300** | 28.5 | 7.4 | 17.95 | **+185%** |

**W&B runs:** `n9ra6pql` (seed 42), `mgeic3k0` (seed 73)

### Analysis

- **Catastrophic failure on seed 42.** Training diverged after epoch 65 — val losses exploded from ~0.87 to >1.6. Best model was captured at epoch 65, but with metrics 4-5x worse than baseline.
- **Seed 73 converged** but still worse than baseline across all metrics (+20% p_in, +5% p_tan, +24% p_oodc, +17% p_re).
- **Root cause hypothesis:** The angle-based sorting from centroid may produce inconsistent/noisy curvature estimates, especially at the trailing edge where nodes from upper and lower surfaces are spatially close but the angle ordering is ambiguous. This noise propagates as a high-variance input feature that destabilizes training.
- The existing `curv` proxy (DSDF gradient magnitude norm) is already a more stable curvature proxy — it's computed from the SDF field which is smooth by construction, whereas finite-difference curvature from discrete mesh nodes amplifies mesh irregularity.
- Seed sensitivity (42 diverges, 73 doesn't) suggests the feature adds noise that sometimes tips training into an unstable basin.

### Suggested follow-ups

1. **Smooth the curvature estimate** — instead of raw finite differences, use a Gaussian-weighted window or fit a local spline before computing curvature. This would reduce mesh-noise sensitivity.
2. **Curvature from DSDF** — compute curvature as ∇²(SDF) at surface nodes using the 8-channel DSDF, which is already smooth. This avoids the angle-sorting problem entirely.
3. **Clamp/normalize curvature values** — the raw κ×chord values may have extreme outliers at LE/TE. Applying asinh or clamp before feeding to the model might stabilize training.
4. **Abandon explicit curvature** — the existing DSDF gradient norm proxy may already encode sufficient curvature information. The model might not need more explicit geometry.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2231 because ADVISOR: Closing — catastrophic failure. Seed 42 diverged at epoch 65 (p_in 47.9, +300% over baseline). Even seed 73, which converged, shows all metrics worse by +5-24%.

Root cause is clear: finite-difference curvature from discrete mesh nodes amplifies mesh irregularity, especially at TE where upper/lower surfaces are close and angle sorting is ambiguous. The resulting high-variance feature destabilizes training (seed 42 divergence confirms this).

Your suggestion to use DSDF-based curvature (∇²SDF) is interesting but at this point, after 7 consecutive feature engineering failures this round, the evidence is overwhelming: **the feature engineering space is saturated.** DSDF + TE coord frame + wake deficit fully capture the useful geometric information. All attempts at additional features (normals, curvature, arc-length, chord-camber, Re-walldist, domain conditioning) have failed.

Shifting you to a completely different direction — loss function reformulation..
